### DISCUSSION TITLE
ECSQL "is null" "is not null"

### DISCUSSION THREAD

**Discussion by Paolo-lab** [iTwin/itwinjs-core#5154] [Q&A]
Hello everyone good evening, I'm having problems in ECSQL with "IS NULL" AND "NULL"
this is the query that I launch in the Imodelconsole
as you can see from the query I'm looking for all the elements in which the "ID__x0020__Cronoprogramma" property is not null
but the query returns me, as you can see from the image, all the elements in which the "ID__x0020__Cronoprogramma" property is null

thanks for the precious help
![Not Null](https://user-images.githubusercontent.com/57446652/222166054-6ceadeaf-69d9-4184-bea8-a6e2e3e46061.JPG)

**Comment by pmconne** [comment]
@khanaffan

**Reply by khanaffan** [reply]
@Paolo-lab Can you give me the following information
1. Count of NULL `SELECT COUNT(*) FROM DgnCustomItemTypes_IT_Pset_Rail.Pset_BallastElementAspect WHERE ID_x0020__Cronoprogramma IS NULL`
2. Count of NOT NULL `SELECT COUNT(*) FROM DgnCustomItemTypes_IT_Pset_Rail.Pset_BallastElementAspect WHERE ID_x0020__Cronoprogramma IS NOT NULL`

Also, try following by aliasing the property `ID__x0020__Cronoprogramma`
 `SELECT ID__x0020__Cronoprogramma AS Cronoprogramma FROM DgnCustomItemTypes_IT_Pset_Rail.Pset_BallastElementAspect WHERE ID_x0020__Cronoprogramma IS NOT NULL`
 
Also, try following by using wild card selection and tell me if you see value for `ID_x0020__Cronoprogramma` 
 `SELECT * FROM DgnCustomItemTypes_IT_Pset_Rail.Pset_BallastElementAspect WHERE ID_x0020__Cronoprogramma IS NOT NULL`

I will be waiting for your feedback.

Thanks

**Reply by candcconsulting** [reply]
Also note @Paolo-lab imodelconsole reports NULL for a blank string '' ... therefore IS NOT NULL AND != '' (two single quotes) should remove the blank records

**Comment by Paolo-lab** [comment]
Hi, sorry if I didn't reply earlier.
thanks for the advice "candconsulting" :-) so it seems to work but I do not understand why. is it a rule? i.e. should null items be searched for in this mode or is it an anomaly?

I made two sceenshoots of the "khanaffan" requests which I am attaching.
as you can see "Is null" and "is not null" are not respected unless I add the suggestion of "candconsulting"

thanks for the support
![provaQuery](https://user-images.githubusercontent.com/57446652/222446985-eeeee85d-f002-402e-9567-bfe626db56ee.JPG)

![provaQuery1](https://user-images.githubusercontent.com/57446652/222447012-f0d02df5-7306-46e1-b00c-ce485636bb77.JPG)

**Reply by khanaffan** [reply]
It's a bug in IModelConsole and created an issue to track and fix it. 
https://github.com/iTwin/itwinjs-core/issues/5161

**Comment by pmconne** [comment]
> imodelconsole reports NULL for a blank string 

That's...confusing. @khanaffan @Josh-Schifter is that intentional?

**Reply by khanaffan** [reply]
I created an Issue for it as it's a bug in IModelConsole https://github.com/iTwin/itwinjs-core/issues/5161

**Comment by Paolo-lab** [comment]
Could it be that the file or the class is corrupted ie a problem arose in creating the digital twin?

**Reply by khanaffan** [reply]
No, it's just a bug in IModelConsole as I see, and nothing is corrupted. The IModel you are using has a property, for instance, set to an empty string.  

@abeesh might not know why connectors bring in an empty string as an empty string and not NULL for `DgnCustomTypes`. It might be a good idea to set empty strings as NULL in IModel.
